Migrate the `delphiBackfillCorrections` package from the `covid-19` repo

nmdefries opened the pull request:
### Description
Move the `covid-19/jingjing/backfill_correction` branch and associated commit history to `covidcast-indicators`.

Used [this approach](https://mattbanderson.com/migrating-a-branch-between-git-repositories-and-preserving-history/) to preserve the commit history. Additional [Slack discussion here](https://delphi-org.slack.com/archives/C013AH5N01E/p1663783958834179).

Additional package setup (e.g. tests on push) will be done in a separate PR.

### Changelog
- Current version of `delphiBackfillCorrections` package.Slack discussion here

### Fixes
Partially addresses #1700

nmdefries commented:
@krivard in case you want to look at this.

krivard commented:
were 94 files expected for this change, or did some auto-generated files slip in by mistake?

nmdefries commented:
All the files look good. There are a large number of documentation-related files (argument templates and `.Rd` function description files), but we need to include those.

krivard commented on Backfill_Correction/delphiBackfillCorrection/R/model.R:
```diff
@@ -0,0 +1,259 @@
+
+    # Save model to cache.
+    create_dir_not_exist(dirname(model_path))
+    save(obj, file=model_path)
```
this might be a better place to record the date trained for this model

jingjtang replied:
We do have the `training_end_date` serving as the prefix of the saved model names.

nmdefries replied:
Addressed in https://github.com/cmu-delphi/covidcast-indicators/pull/1711

krivard commented on Backfill_Correction/delphiBackfillCorrection/R/preprocessing.R:
```diff
@@ -0,0 +1,222 @@
+## Data Preprocessing
+## 
+## The raw input data should have 4/5 basic columns:
```
for files with only 4 columns, which columns are optional?

jingjtang replied:
`issue date` and `lag`, we should have at least one of it, since the other one can be computed accordingly given `time_value`

jingjtang replied:
So, the 4 columns would be `time_value`, `lag`/`issue_date`, `denom`, `num`

krivard commented on Backfill_Correction/delphiBackfillCorrection/R/preprocessing.R:
```diff
@@ -0,0 +1,222 @@
+
+#' Get week of a month info according to a date
+#'
+#' All the dates on or before the ith Sunday but after the (i-1)th Sunday
+#' is considered to be the ith week. Notice that the dates in the 5th week
+#' this month are actually in the same week with the dates in the 1st week
+#' next month and those dates are sparse. Thus, we assign the dates in the 
+#' 5th week to the 1st week.
```
So are October 2022 weeks numbered like this:

```
𝙾𝙲𝚃𝙾𝙱𝙴𝚁 – 𝟸𝟶𝟸𝟸
𝚂𝚞|𝙼𝚘|𝚃𝚞|𝚆𝚎|𝚃𝚑|𝙵𝚛|𝚂𝚊
------------------
𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝟶𝟷 week 1
𝟶𝟸|𝟶𝟹|𝟶𝟺|𝟶𝟻|𝟶𝟼|𝟶𝟽|𝟶𝟾 week 2
𝟶𝟿|𝟷𝟶|𝟷𝟷|𝟷𝟸|𝟷𝟹|𝟷𝟺|𝟷𝟻 week 3
𝟷𝟼|𝟷𝟽|𝟷𝟾|𝟷𝟿|𝟸𝟶|𝟸𝟷|𝟸𝟸 week 4
𝟸𝟹|𝟸𝟺|𝟸𝟻|𝟸𝟼|𝟸𝟽|𝟸𝟾|𝟸𝟿 week 1 November
𝟹𝟶|𝟹𝟷|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡 week 2 November
```

?

jingjtang replied:
No, it should be
```
𝙾𝙲𝚃𝙾𝙱𝙴𝚁 – 𝟸𝟶𝟸𝟸
𝚂𝚞|𝙼𝚘|𝚃𝚞|𝚆𝚎|𝚃𝚑|𝙵𝚛|𝚂𝚊
------------------
𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝟶𝟷 week 1
𝟶𝟸|𝟶𝟹|𝟶𝟺|𝟶𝟻|𝟶𝟼|𝟶𝟽|𝟶𝟾 week 1
𝟶𝟿|𝟷𝟶|𝟷𝟷|𝟷𝟸|𝟷𝟹|𝟷𝟺|𝟷𝟻 week 2
𝟷𝟼|𝟷𝟽|𝟷𝟾|𝟷𝟿|𝟸𝟶|𝟸𝟷|𝟸𝟸 week 3
𝟸𝟹|𝟸𝟺|𝟸𝟻|𝟸𝟼|𝟸𝟽|𝟸𝟾|𝟸𝟿 week 4
𝟹𝟶|𝟹𝟷|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡 week 1 November
```
Sorry this function is not up-to-date, updated just now.

krivard reviewed (commented):
Comments up through tooling.R; I'll come back and do the rest later unless you tell me I'm trying too hard for this review 😂 

Additionally:
* None of our other toplevel directories include upper case. How strongly do you feel about the casing in the directory name `Backfill_Correction`?

nmdefries commented:
> How strongly do you feel about the casing in the directory name

I would prefer to rename it to match the format of the other packages here; this is just an as-is copy of the package.

How about naming it `backfill_corrections`?

jingjtang commented on Backfill_Correction/delphiBackfillCorrection/R/main.R:
```diff
@@ -0,0 +1,273 @@
+    file.remove(file.path(mydir, files_list))
+  }
+    
+  training_end_date <- as.Date(readLines(
+    file.path(params$cache_dir, "training_end_date.txt")))
```
It's super easy to save and read the `training_end_date` this way. The only con is that we will have an extra txt file. 

We have the `training_end_date` stored as `prefix` of the saved model files as well. We can also get the date there.

nmdefries replied:
`training_end_date` will automatically change over time as we train new models. Updating params automatically in tandem with that would be finicky and require more complicated `make` logic. I think it's simpler to work off of a local cache, especially since we're already caching other objects.

We can support an optional `training_end_date` setting in `params` if we think that's important. I suppose if it were set, it would supersede the date derived from the cache.

krivard replied:
what does `training_end_date` get set to if there's no model yet?

nmdefries replied:
Addressed in https://github.com/cmu-delphi/covidcast-indicators/pull/1711

jingjtang commented on Backfill_Correction/delphiBackfillCorrection/R/model.R:
```diff
@@ -0,0 +1,259 @@
+#' @export
+add_sqrtscale<- function(train_data, test_data, max_raw, value_col) {
+  if (!(value_col %in% colnames(train_data))){
+    stop("value raw does not exist in training data!")
```
This checking is design more for a tooling usage. In our daily production, it won't happen.

jingjtang commented on Backfill_Correction/delphiBackfillCorrection/R/model.R:
```diff
@@ -0,0 +1,259 @@
+      error=function(e) {print(paste("Training failed for", model_path, sep=" "))}
+    )
+  }
+  if (success < 9) {return (NULL)}
```
I initially design it to include as detailed information as it can (especially for tooling usage). And if a model training is fail, it's super fast, won't take much time. 

But you are right, here we can stop once there is a failure

nmdefries replied:
9 is the length of the default `taus` list. Switched this to actually checking the length of the list.

jingjtang commented on Backfill_Correction/README.md:
```diff
@@ -0,0 +1,122 @@
+   one signal at the state level, a model takes about 30 minutes to train. Due
+   to resource limitations in production, we only train models once a month
+   and save the model objects between runs. By default, these are saved to the
+   `cache` directory.
```
Updated. Is the current one better?

jingjtang commented on Backfill_Correction/README.md:
```diff
@@ -0,0 +1,122 @@
+- lag: the number of days between issue date and the reference date
+- issue_date: issue date/report, required if lag is not available
+
+Required columns without fixed column names:
```
Usually we have `denom` for the denominator (number of total counts) and `num` for the numerator(number of covid/interested counts). Quidel is a little bit different. We have more columns with different suffixes to indicate age groups.

nmdefries commented on Backfill_Correction/delphiBackfillCorrection/R/delphiBackfillCorrection.R:
```diff
@@ -0,0 +1,4 @@
+# Suppress R CMD check note
+#' @importFrom tibble tribble
+#' @import covidcast
+NULL
```
This loads packages/function that are used outside of `delphiBackfillCorrection`-defined functions, e.g. setting global variables, and can't be documented in the normal way.

krivard reviewed (commented):
some followups, but also @nmdefries something has gone a bit weird with the file organization here -- there are files that used to be in delphiBackfillCorrection that are now only found deep in the delphiBackfillCorrection.Rcheck directory. What is supposed to be here?

nmdefries commented:
Whoops, I added local test output files when renaming the directory. Removed in https://github.com/cmu-delphi/covidcast-indicators/pull/1701/commits/fb38e07ac9b292a070bc958a1cd183da6cda33b1

nmdefries commented on backfill_corrections/delphiBackfillCorrection/R/main.R:
```diff
@@ -0,0 +1,273 @@
+  if (params$train_models) {
+    # Remove all the stored models
+    files_list <- list.files(params$cache_dir, pattern="*.model", full.names = TRUE)
+    file.remove(file.path(mydir, files_list))
```
```suggestion
    file.remove(files_list)
```
@jingjtang Since `files_list` already has the full paths, I don't think we need `file.path()` in the remove call.

jingjtang commented:
> are there any constraints on the name, like does it have to fit a certain pattern? otherwise how does the code know which columns do what?

@krivard There would be a constraint when the functions are used for tooling. 

But here, for the daily production, the uncertainty is created by Quidel since Quidel has multiple age group signals. We don't want to add burden on the storage of backfill files. So, currently, for all the other indicators, `num` and `denom` are the two fixed columns, but we have some extra ones like `num_age_0_4` etc. for Quidel.

krivard commented on Backfill_Correction/delphiBackfillCorrection/R/beta_prior_estimation.R:
```diff
@@ -0,0 +1,212 @@
+  train_data$value_7dav = NaN
+  train_data$value_prev_7dav = NaN
+  
+  #For testing
```
I think I'm missing context here but resolving that can wait until this project is farther along

krivard reviewed (approved):
I believe this is a complete pass, and I'm provisionally fine with merging this as-is or with the suggested changes.

I do want to flag a couple of things for later discussion and possible refactor:
* keeping track of models and their training end dates
* flow of training and test data through prediction and training modes

jingjtang commented on backfill_corrections/delphiBackfillCorrection/man-roxygen/testing_window-template.R:
```diff
@@ -0,0 +1,2 @@
+#' @param testing_window the testing window used for saving the runtime. Could
+#'     set it to be 1 if time allows
```
I would suggest to train it every 2 weeks or every 4 weeks

jingjtang replied:
Since we run it daily, we can just set it to 1. If for some reasons we don't have the prediction result for several days and don't traverse the training date, then we can produce the prediction for those days at once by change the value of testing window

jingjtang replied:
To what ever number of days of prediction we missed.

jingjtang commented on backfill_corrections/delphiBackfillCorrection/R/main.R:
```diff
@@ -0,0 +1,273 @@
+      indicator = input_group$indicator, signal = input_group$signal,
+      signal_suffixes = input_group$name_suffix)
+    
+    if (params$train_models) {
+      # Save the training end date to a text file.
+      writeLines(as.character(TODAY), 
+                 file.path(params$cache_dir, "training_end_date.txt"))
+    }
```
Both of them are controlled by the same variable `params$train_models`. If it is TRUE, then both the models and the `traing_end_date.txt` will be updated

nmdefries replied:
Addressed in https://github.com/cmu-delphi/covidcast-indicators/pull/1711

nmdefries commented:
@krivard This is ready to merge.

Additional package setup (e.g. tests on push) will be done in a separate PR. Expect some changes after the planning meeting.
